docs/DesignDocs/TEECallsRefactoring.md

```diff
@@ -300,218 +282,397 @@ else
+of the three imported functions. Assuming that we use *weak symbol* to avoid the
+duplication of the wrapper functions (i.e., only one implementation of the functions is picked),
+the invocation of these ECALLs to both enclaves ends up using the same function ids.
 
-Each registration function in a EDL file is added to a list via the constructor.
-On Linux, we can do this via ` __attribute__((constructor))` as follows.
-We can use an equivalent approach for the constructor on Windows.
+See the following host-side code snippet for an exmaple.
 ```
-/* Constructor to register the callback. */
- __attribute__((constructor)) static void core_ecall_functions_constructor()
-{
-    oe_ecall_init(register_core_ecall_functions);
-}
+// ECALL to the enclave `foo`.
+common_1_ecall(enclave_foo);
+
+// ECALL to the enclave `bar`.
+common_1_ecall(enclave_bar);
 ```
 
-The list of registration functions can then be invoked during the host and enclave
-initialization.
-  
-**Cache the Function Id for the Const-Time Look-Up**
+Since the host can have only one implementation of `ecall_common_1_ecall` (the wrapper function),
+both invocations end up using the same function id, say `0`. However, this causes the mismatch
+on the call into the enclave `bar` where the expected id should be `2`.
 
-One important property of current implementation is the const-time function look-up that
-this proposal would like to maintain. To this end, the design is obtaining the id by hash
-in the first invocation and caching the id for subsequent ones.
+Proposed revision
+-----------
 
-- Caching an OCALL id.
+The main goal of this proposal is to address the limitation that different enclaves on the
+same host cannot import the same EDL files (i.e., uinsg the same set of ECALLs). In addition,
+the proposed revision aims to remove the needs of `_u_wrapper.c` and `_t_wrapper.c`,
+which use uncommon practices (i.e., include the source file).
 
-  Because an enclave only interacts with a single host, the case of OCALL wrapper is straightforward:
-  caching the id within the implementation of the wrapper.
+To meet these goals, this document proposes a two-phase revision.
+Each phase requires separate reviews and implementation.
 
+## Phase 1: Revise the ECALL Calling Mechanism
+
+**Store ECALL by hash**
+
+The main drawback of current implementation is using the table defined
+in the `_u.c/_t.c` that serves as the contract between both ends.
+Given that the rule of importing `EDL` files is flexible, the resulting
+ECALL table is not determinstic (i.e., can be any order based on how
+EDL files are imported).
+
+To solve the problem, this phase proposes to using a single table
+that stores the function pointer along with its hash (e.g., the hash of the function name).
+Take the previous example with two enclaves, the ECALL tables now beome:
+
+- `foo` ECALL table
+  ```
+  oe_ecall_struct_t __oe_ecalls_table[] = {
```

Comment:
As stated above, the problem happens only on the host side. 
> Since the host can have only one implementation of `ecall_common_1_ecall` (the wrapper function),
initialization. both invocations end up using the same function id, say `0`. However, this causes the mismatch on the call into the enclave `bar` where the expected id should be `2`.

Therefore it is not fully justified why the enclave side table needs to change.

Consider this scheme  (not optimized; but a good starting point) where the mapping between the global ecall function identifier (say the function name) to the local function identifier (the ecall table index) is done entirey on the hsot side:
```
   1. Each oe_enclave_t maintains an array of ecall names.
   2. To invoke an ecall, the function name is passed in.
   3. Based on the function name, the appropriate index into the ecall table can be found out
   4. Use the index (local id) to dispatch the ecall.
```
This shows that the mapping can be done entirely on the host side. It then becomes a matter of doing the mapping efficiently.

Reply:
Another goal of this design is ultimately to remove the needs of  `_t_wrapper.c`/`_u_wrapper.c`; that is, the enclave don't need to maintain multiple ECALL tables and to have host passing the hard-coded table id.  However, if the end goal is making all the internal EDL files optional (may the case for OCALLs, but how about ECALLs?), I'd agree that we only need host-side modifications.

Another advantage of using hash is its potential of extending it to do the runtime checks (e.g., type checks), which cannot be done by simply passing id.

Other than the those, the host side mapping can surely can done similarly to this proposal and provides the same efficiency. Another advantage is removing the need of the special ECALL for querying the ECALL id.

Reply:
I've updated the proposal to scope down and only make host-side modifications.